hack/generator/pkg/astmodel/armconversion/convert_to_arm_function_builder.go

```diff
@@ -184,6 +185,134 @@ func (builder *convertToARMBuilder) referencePropertyHandler(
+	if err != nil {
+		panic(err)
+	}
+
+	needToInitializeToProp := false // we need to init the target if it is optional
+	var toPropTypeName astmodel.TypeName
+	// (2.)  resolve any optional type
+	if optType, ok := toPropType.(*astmodel.OptionalType); ok {
+		needToInitializeToProp = true
+		// (3.) resolve any inner typename
+		toPropTypeName = optType.Element().(astmodel.TypeName)
+		toPropType, err = allTypes.FullyResolve(optType.Element())
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	// (4.) we have found the underlying object type
+	toPropObjType := toPropType.(*astmodel.ObjectType)
+
+	// *** Now generate the code! ***
+
+	// Build the initializer for the to-prop (if needed)
+	var result []dst.Stmt
+	if needToInitializeToProp {
+		result = []dst.Stmt{builder.buildToPropInitializer(fromProps, toPropTypeName, toPropName)}
+	}
+
+	// Copy each from-prop into the to-prop
+	for _, fromProp := range fromProps {
+		// find the corresponding inner property on the to-prop type
+		toSubProp, ok := toPropObjType.Property(fromProp.PropertyName())
+		if !ok {
+			panic("unable to find source of flattened property")
+		}
+
+		// generate conversion
+		stmts := builder.typeConversionBuilder.BuildConversion(astmodel.ConversionParameters{
+			Source:            astbuilder.Selector(dst.NewIdent(builder.receiverIdent), string(fromProp.PropertyName())),
+			SourceType:        fromProp.PropertyType(),
+			Destination:       astbuilder.Selector(astbuilder.Selector(dst.NewIdent(builder.resultIdent), string(toPropName)), string(toSubProp.PropertyName())),
+			DestinationType:   toSubProp.PropertyType(),
+			NameHint:          string(toSubProp.PropertyName()),
+			ConversionContext: nil,
+			AssignmentHandler: nil,
+			Locals:            builder.locals,
+		})
+
+		// failed?
+		if len(stmts) == 0 {
+			return nil
+		}
+
+		result = append(result, stmts...)
+	}
+
+	return result
+}
+
+func (builder *convertToARMBuilder) buildToPropInitializer(fromProps []*astmodel.PropertyDefinition, toPropTypeName astmodel.TypeName, toPropName astmodel.PropertyName) dst.Stmt {
+	// build (x != nil, y != nil, …)
+	conds := make([]dst.Expr, 0, len(fromProps))
+	for _, prop := range fromProps {
+		conds = append(conds, &dst.BinaryExpr{
+			X:  astbuilder.Selector(dst.NewIdent(builder.receiverIdent), string(prop.PropertyName())),
+			Op: token.NEQ,
+			Y:  dst.NewIdent("nil"),
+		})
+	}
+
+	// build (x || y || …)
+	cond := conds[0]
+	for _, c := range conds[1:] {
+		cond = &dst.BinaryExpr{
+			X:  cond,
+			Op: token.LOR,
+			Y:  c,
+		}
+	}
+
+	// build if (conds…) { target.prop = &TargetType{} }
+	return &dst.IfStmt{
+		Cond: cond,
+		Body: &dst.BlockStmt{
+			List: []dst.Stmt{
+				astbuilder.SimpleAssignment(
+					astbuilder.Selector(dst.NewIdent(builder.resultIdent), string(toPropName)),
+					token.ASSIGN,
+					&dst.UnaryExpr{
+						Op: token.AND,
+						X: &dst.CompositeLit{
+							Type: toPropTypeName.AsType(builder.codeGenerationContext),
+						},
+					}),
```

Comment:
I've got an `astbuilder` helper for `CompositeLit` in one of my PRs, you may want to steal it.

Reply:
I'm not sure it would shorten much here?